packages/amplify-graphql-model-transformer/src/graphql-model-transformer.ts

```diff
@@ -319,22 +319,47 @@ export class ModelTransformer extends TransformerModelBase implements Transforme
         }
       }
       // global auth check
-      if (!hasAuth && ctx.transformParameters.sandboxModeEnabled && ctx.authConfig.defaultAuthentication.authenticationType !== 'API_KEY') {
-        const apiKeyDirArray = [makeDirective(API_KEY_DIRECTIVE, [])];
-        extendTypeWithDirectives(ctx, def.name.value, apiKeyDirArray);
-        propagateApiKeyToNestedTypes(ctx as TransformerContextProvider, def, new Set<string>());
-        queryFields.forEach((operationField) => {
-          const operationName = operationField.name.value;
-          addDirectivesToOperation(ctx, ctx.output.getQueryTypeName()!, operationName, apiKeyDirArray);
-        });
-        mutationFields.forEach((operationField) => {
-          const operationName = operationField.name.value;
-          addDirectivesToOperation(ctx, ctx.output.getMutationTypeName()!, operationName, apiKeyDirArray);
-        });
-        subscriptionsFields.forEach((operationField) => {
-          const operationName = operationField.name.value;
-          addDirectivesToOperation(ctx, ctx.output.getSubscriptionTypeName()!, operationName, apiKeyDirArray);
-        });
+      if (!hasAuth) {
+        const serviceDirectiveNames = new Set<string>();
+
+        const defaultServiceDirectiveName = APPSYNC_AUTH_CONFIGURATION_TYPE_TO_SERVICE_DIRECTIVE_MAP.get(
+          ctx.authConfig.defaultAuthentication.authenticationType,
+        );
+        if (ctx.transformParameters.sandboxModeEnabled && ctx.synthParameters.enableIamAccess) {
+          serviceDirectiveNames.add(API_KEY_DIRECTIVE);
+          serviceDirectiveNames.add(AWS_IAM_DIRECTIVE);
+          if (defaultServiceDirectiveName) {
+            serviceDirectiveNames.add(defaultServiceDirectiveName);
+          }
+        } else if (ctx.transformParameters.sandboxModeEnabled && ctx.authConfig.defaultAuthentication.authenticationType !== 'API_KEY') {
+          serviceDirectiveNames.add(API_KEY_DIRECTIVE);
+          if (defaultServiceDirectiveName) {
+            serviceDirectiveNames.add(defaultServiceDirectiveName);
+          }
+        } else if (ctx.synthParameters.enableIamAccess && ctx.authConfig.defaultAuthentication.authenticationType !== 'AWS_IAM') {
+          serviceDirectiveNames.add(AWS_IAM_DIRECTIVE);
+          if (defaultServiceDirectiveName) {
+            serviceDirectiveNames.add(defaultServiceDirectiveName);
+          }
+        }
```

Comment:
todo: remove default fallback.
if there's no `@auth` we should just check sandbox and iam check.

Reply:
this is fixed by https://github.com/aws-amplify/amplify-category-api/pull/2331/commits/5935803b0aea7f48194cd32aa4534f70c7a353d9